src/TreeSitter/Importing.hs

```diff
@@ -35,54 +36,57 @@ import           TreeSitter.Node as TS
 import           TreeSitter.Parser as TS
 import           TreeSitter.Tree as TS
 import           Data.Proxy
+import           Prelude hiding (fail)
+import           Type.Reflection
 
 -- Parse source code and produce AST
-parseByteString :: Building t => Ptr TS.Language -> ByteString -> IO (Maybe t)
+parseByteString :: Building symbol t => Ptr TS.Language -> ByteString -> IO (Either String t)
 parseByteString language bytestring = withParser language $ \ parser -> withParseTree parser bytestring $ \ treePtr ->
   if treePtr == nullPtr then
-    pure Nothing
+    pure (Left "error: didn't get a root node")
   else
     withRootNode treePtr $ \ rootPtr ->
       withCursor (castPtr rootPtr) $ \ cursor ->
-        runMaybeC (runM (runReader cursor (runReader bytestring buildNode)))
+        runM (runFail (runReader cursor (runReader bytestring buildNode)))
 
 -- | Building is the process of iterating over tree-sitter’s parse trees using its tree cursor API and producing Haskell ASTs for the relevant nodes.
 --
 --   Datatypes which can be constructed from tree-sitter parse trees may use the default definition of 'buildNode' providing that they have a suitable 'Generic' instance.
-class Building a where
-  buildNode :: (Alternative m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m a
-  default buildNode :: (Alternative m, Carrier sig m, GBuilding (Rep a), Generic a, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m a
+class Building symbol a | a -> symbol where
+  buildNode :: (MonadFail m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m a
+  default buildNode :: (MonadFail m, Carrier sig m, GBuilding (Rep a), Generic a, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m a
   buildNode = to <$> gbuildNode
 
-  buildEmpty :: Alternative m => m a
-  buildEmpty = empty
-
-instance Building Text.Text where
-  buildNode = do
-    node <- peekNode
-    case node of
-      Just node' -> do
-        bytestring <- ask
-        let start = fromIntegral (nodeStartByte node')
-            end = fromIntegral (nodeEndByte node')
-        pure (decodeUtf8 (slice start end bytestring))
-      _ -> empty
+  buildEmpty :: MonadFail m => m a
+  buildEmpty = fail "expected a node" -- TODO: log what type actually expected the node / what `a` is
 
-instance Building a => Building (Maybe a) where
+instance Building symbol a => Building symbol (Maybe a) where
```

Comment:
Were we still planning on backing the introduction of the `symbol` parameter out?

Reply:
Yes! Just wanted to push these changes to have breadcrumbs for myself and know what we've tried.